### PR TITLE
Refactor routing guards and test them

### PR DESCRIPTION
**Pull request by lderrickable**
refactor routing guards into resolvers and effects

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
Routing guards aren't dry, aren't tested, and aren't all properly categorized as "guards".

## What is the new behavior?
Routing guards are better written and are mostly tested. The CartLoadGuard will become a CartLoadResolver service: https://github.com/graycoreio/daffodil/issues/205

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: apps/foundation-demo/src/app/helper/routing-guards/cart-load.guard.spec.ts [hunk 1]
```diff
@@ -35,43 +36,48 @@ xdescribe('CartLoadGuard', () => {
...
     stubCart = cartFactory.create();
     store = TestBed.get(Store);
+    spyOn(store, 'dispatch');
   }));
 
   it('should be created', () => {
     expect(cartLoadGuard).toBeTruthy();
   });
 
-  describe('when the cart is null', () => {
```
Comment (damienwebdev) [inline apps/foundation-demo/src/app/helper/routing-guards/cart-load.guard.spec.ts]: Why did the tests have to change?

Comment (lderrickable) [reply apps/foundation-demo/src/app/helper/routing-guards/cart-load.guard.spec.ts]: I didn't think `CartLoad` was _actually_ a guard, so I removed the `CanActivate` function. It doesn't decide if the route Activatable. It just gets the Cart. So `CartLoadGuard` is really more of a helper function file.

Comment (lderrickable) [reply apps/foundation-demo/src/app/helper/routing-guards/cart-load.guard.spec.ts]: This isn't relevant anymore. closing

File: apps/demo/src/app/routing-resolvers/cart/effects/cart-resolver.effects.spec.ts [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    let expected;
+    
+    describe('and action.payload is null', () => {
+      const resolveCartAction = new ResolveCart(null);
+      const resolveCartFailAction = new ResolveCartFail(null);
+      
+      beforeEach(() => {
+        spyOn(router, 'navigateByUrl');
+        actions$ = hot('--a', { a: resolveCartAction });
+        expected = cold('--b', { b: resolveCartAction, c: resolveCartFailAction });
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply apps/demo/src/app/routing-resolvers/cart/effects/cart-resolver.effects.spec.ts]: `resolveCartFailAction` is the second kind of action that the effect can receive. I don't completely understand how this test works actually. I know that the effect doesn't actually run without the first expect, and that the first expect doesn't pass without `c`. I also know that it does test what I'm trying to test.

File: apps/demo/src/app/routing-resolvers/cart/effects/cart-resolver.effects.spec.ts [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+    describe('and action.payload is defined', () => {
+      
+      beforeEach(() => {
+        const resolveCartAction = new ResolveCart(stubCart);
+        const resolveCartFailAction = new ResolveCartFail(null);
+        spyOn(router, 'navigateByUrl');
+
+        actions$ = hot('--a', { a: resolveCartAction });
+        expected = cold('--b', { b: resolveCartAction, c: resolveCartFailAction });
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply apps/demo/src/app/routing-resolvers/cart/effects/cart-resolver.effects.spec.ts]: Yes, see above comment

File: apps/demo/src/app/routing-resolvers/cart/resolvers/cart-resolver.service.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+@Injectable({
+  providedIn: 'root'
+})
+export class CartResolver implements Resolve<string> {
+  constructor(
+    private store: Store<fromCart.State>
+  ) {}
+
+  resolve(): string {
+    this.store.dispatch(new ResolveStart());
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply apps/demo/src/app/routing-resolvers/cart/resolvers/cart-resolver.service.ts]: I was trying to keep effect-like things in effects, but sure. I can just watch the dispatch stream here instead

File: apps/demo/src/app/routing-resolvers/cart/actions/cart-resolver.actions.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import { Action } from '@ngrx/store';
+import { Cart } from '@daffodil/core';
+
+export enum CartResolverActionTypes {
+  ResolveStartAction = "[Cart-Resolver] Resolve Start Action",
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply apps/demo/src/app/routing-resolvers/cart/actions/cart-resolver.actions.ts]: changed

File: apps/demo/src/app/helper/routing-resolvers/cart/actions/cart-resolver.actions.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import { Action } from '@ngrx/store';
+import { Cart } from '@daffodil/core';
+
+export enum CartResolverActionTypes {
+  ResolveStartAction = "[Cart-Resolver] Resolve Start Action",
+  ResolveCartAction = "[Cart-Resolver] Resolve Cart Action",
+  ResolveCartFailAction = "[Cart-Resolver] Resolve Cart Fail Action"
+}
+
+export class ResolveStart implements Action {
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply apps/demo/src/app/helper/routing-resolvers/cart/actions/cart-resolver.actions.ts]: changed to ResolveCart

File: apps/demo/src/app/routing-resolvers/cart/effects/cart-resolver.effects.spec.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (lderrickable) [reply apps/demo/src/app/routing-resolvers/cart/effects/cart-resolver.effects.spec.ts]: this logic was moved into the cart-resolver, so it isn't tested this way anymore

File: apps/demo/src/app/routing-resolvers/cart/effects/cart-resolver.effects.spec.ts [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+  it('should be created', () => {
+    expect(effects).toBeTruthy();
+  });
+
+  describe('when ResolveCartAction is triggered', () => {
+
+    let expected;
+    
+    describe('and action.payload is null', () => {
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply apps/demo/src/app/routing-resolvers/cart/effects/cart-resolver.effects.spec.ts]: Moved this logic into the CartResolver and rewrote the tests

[the remark this answers is not in the document]
Comment (lderrickable) [reply apps/demo/src/app/routing-resolvers/cart/effects/cart-resolver.effects.spec.ts]: same as above

File: apps/demo/src/app/routing-resolvers/cart/effects/cart-resolver.effects.spec.ts [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+      });
+      
+      it('should navigate to /cart', () => {
+        expect(effects.onResolveCart$).toBeObservable(expected);
+
+        expect(router.navigateByUrl).toHaveBeenCalledWith('/cart');
+      });
+    });
+
+    describe('and action.payload is defined', () => {
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply apps/demo/src/app/routing-resolvers/cart/effects/cart-resolver.effects.spec.ts]: same as above

File: apps/demo/src/app/routing-resolvers/cart/resolvers/cart-resolver.service.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (lderrickable) [reply apps/demo/src/app/routing-resolvers/cart/resolvers/cart-resolver.service.ts]: Changed

File: package-lock.json [hunk 9]
```diff
@@ -3038,7 +3038,7 @@
     },
     "ansi-escapes": {
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply package-lock.json]: Yeah, i committed this by accident. I'll remove it when I squash

File: apps/demo/src/app/routing-resolvers/cart/effects/cart-resolver.effects.spec.ts [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    });
+  });
+
+
+  describe('when CartResolverActionTypes.ResolveStartAction is triggered', () => {
+    
+    let expected;
+    const resolveStartAction = new ResolveStart();
+    
+    describe('and cart in store is defined', () => {
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply apps/demo/src/app/routing-resolvers/cart/effects/cart-resolver.effects.spec.ts]: I tried to clarify it

**Comment by lderrickable** [comment]
@damienwebdev I had some trouble with getting the router to continue to the checkout page after the resolvers resolved. They're definitely resolving and everything, and the only difference between the resolvers now compared to how I had it working before is that the resolvers are returning `Observable<string>` instead of just `string`. If you get time, could you take a quick look and see if it's something small I'm missing?

**Comment by lderrickable** [comment]
squashed and removed the package-lock.json file

**Pull request merged**